src/feature_extraction.py

```diff
@@ -425,25 +425,126 @@ def _dynamic_features(self, credit_card, **kwargs):
 
 
 class POSCASHBalanceFeatures(BasicHandCraftedFeatures):
+    def __init__(self, last_k_agg_periods, last_k_trend_periods, num_workers=1, **kwargs):
+        super().__init__(num_workers=num_workers)
+        self.last_k_agg_periods = last_k_agg_periods
+        self.last_k_trend_periods = last_k_trend_periods
+
+        self.num_workers = num_workers
+        self.features = None
+
     def fit(self, pos_cash, **kwargs):
         pos_cash['is_contract_status_completed'] = pos_cash['NAME_CONTRACT_STATUS'] == 'Completed'
-        pos_cash_sorted = pos_cash.sort_values(['SK_ID_CURR', 'MONTHS_BALANCE'])
+        pos_cash['pos_cash_paid_late'] = (pos_cash['SK_DPD'] > 0).astype(int)
+        pos_cash['pos_cash_paid_late_with_tolerance'] = (pos_cash['SK_DPD_DEF'] > 0).astype(int)
 
         features = pd.DataFrame({'SK_ID_CURR': pos_cash['SK_ID_CURR'].unique()})
-
-        g = pos_cash_sorted.groupby('SK_ID_CURR')['CNT_INSTALMENT_FUTURE'].last().reset_index()
-        g.rename(index=str,
-                 columns={'CNT_INSTALMENT_FUTURE': 'pos_cash_remaining_installments'},
-                 inplace=True)
-        features = features.merge(g, on=['SK_ID_CURR'], how='left')
-
         groupby = pos_cash.groupby(['SK_ID_CURR'])
-        g = groupby['is_contract_status_completed'].agg('sum').reset_index()
-        features = features.merge(g, on=['SK_ID_CURR'], how='left')
+        func = partial(POSCASHBalanceFeatures.generate_features,
+                       agg_periods=self.last_k_agg_periods,
+                       trend_periods=self.last_k_trend_periods)
+        g = parallel_apply(groupby, func, index_name='SK_ID_CURR', num_workers=self.num_workers).reset_index()
+        features = features.merge(g, on='SK_ID_CURR', how='left')
 
         self.features = features
         return self
 
+    @staticmethod
+    def generate_features(gr, agg_periods, trend_periods):
+        one_time = POSCASHBalanceFeatures.one_time_features(gr)
+        all = POSCASHBalanceFeatures.all_installment_features(gr)
+        agg = POSCASHBalanceFeatures.last_k_installment_features(gr, agg_periods)
+        trend = POSCASHBalanceFeatures.trend_in_last_k_installment_features(gr, trend_periods)
+        last = POSCASHBalanceFeatures.last_loan_features(gr)
+        features = {**one_time, **all, **agg, **trend, **last}
+        return pd.Series(features)
+
+    @staticmethod
+    def one_time_features(gr):
+        gr_ = gr.copy()
+        gr_.sort_values(['MONTHS_BALANCE'], inplace=True)
+        features = {}
+
+        features['pos_cash_remaining_installments'] = gr_['CNT_INSTALMENT_FUTURE'].tail(1)
+        features['pos_cash_completed_contracts'] = gr_['is_contract_status_completed'].agg('sum')
+
+        return features
+
+    @staticmethod
+    def all_installment_features(gr):
```

Comment:
@kstrzala is the naming just an artifact or conscious decision

Reply:
I know it may seem as an artifact (from copy/paste installments features), but because in this table we also deal with installments, I left the names consciously.

Reply:
ok, good.